Implement dynamic watch manager responsible for event dispatch to watching controllers

shomron opened the pull request:
Implement dynamic watch manager responsible for event dispatch to watching controllers

This commit is significant change in how dynamic watches are implemented and routed. Previously, the watch manager would
manage the lifecycle of subordinate controllers via a subordinate controller-runtime Manager which was rebuilt whenever
the set of watched resources would change. This would dispose of informer caches and require a full resync from the API server.
This approach worked around the inability of controller-runtime to remove active informers and controllers running within a Manager.

Here we introduce a new approach where there is only a single controller-runtime Manager and controllers are fixed.
Instead, the informer cache has been forked to allow removal of informers once they are no longer needed.
Since controllers still cannot be removed and event handlers cannot be detached from a running informer, all events for
dynamically-watched resources are routed through the watch manager and distributed to the interested parties.
There is now a single constraint controller for all constraint kinds, and a single sync controller for all synced resources.

* Fork controller-runtime 0.5.0's cache.Cache into a new DynamicCache, supporting removal of informers from its informer map.
* Introduce non-blocking API to DynamicCache to support adding informers without waiting for them to sync.
* Rewrite watch.Manager to add/remove informers in the dynamicCache. Informer events will be sent to the watch manager,
  and distributed to any watching registrars.
* Constraint and Sync controllers are now singletons which demultiplex incoming events and process multiple resource kinds.
* Add tests for the config controller verifying cache correctness and interactions with config set changes.
* Sync controller will now drop events for resources that are no longer in the active watch set.
* Removed ordering constraints between ConstraintTemplate and Constraint controllers - deletion can be processed out-of-order.

Signed-off-by: Oren Shomron <shomron@gmail.com>

```
┌────────────────────────────────────────────┐           ┌─────────────────────────────────────┐
│               watch.Manager                │           │           Informer Cache            │
│                                            │           │                                     │
│                    ┌────────────────┐      │           │                                     │
│┌──────┬──Fan-out───│  EventHandler  │◀─────┼─────┐     │┌───────────────────────────────────┐│
││      │            └────────────────┘      │     ├─────┤│             Informer              ││
││      │                                    │     │     │└───────────────────────────────────┘│
││      ▼                                    │     │     │┌───────────────────────────────────┐│
││  ┌──────┐    ┌──────────────────────────┐ │     ├─────┤│             Informer              ││
││  │Events├ ─ ─│Registrar (constraint-ctl)│ │     │     │└───────────────────────────────────┘│
││  └──────┘    └──────────────────────────┘ │     │     │┌───────────────────────────────────┐│
││  ┌──────┐    ┌──────────────────────────┐ │     └─────┤│             Informer              ││
│└─▶│Events├ ─ ─│Registrar (config-ctl)    │ │           │└───────────────────────────────────┘│
│   └───┬──┘    └──────────────────────────┘ │           │                                     │
└───────┼────────────────────────────────────┘           └─────────────────────────────────────┘
        │
        │
        │ source.Channel
        │
        │
        │   ┌─────────────────────┐
        │   │Constraint Controller│
        │   │      (Single)       │
        │   │                     │
        └──▶│                     │
            │                     │
            │                     │
            └─────────────────────┘
```
Relates to discussion in #473

maxsmythe commented on main.go:
```diff
@@ -84,6 +85,7 @@ func main() {
 	ctrl.SetLogger(crzap.Logger(true))
 
 	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		NewCache:               dynamiccache.New,
```
If the two managers share the same cache, the parent manager needs to have some way of making sure child managers don't un-watch resources that the parent manager is interested in (e.g. Constraint Templates here).

shomron replied:
I believe that is handled [here](https://github.com/open-policy-agent/gatekeeper/pull/490/files#diff-17f381e8f516b395e2a381a78503681eR162-R165)?

maxsmythe replied:
The link is broken for me. Where is this handled? I kept an eye out, but couldn't find anything obvious.

shomron replied:
Let me make sure I understand the question.
Within a watch manger, with multiple registrars watching the same resource, the watch manager performs reference counting and only removes the informer when the last registrar unregisters. That happens in this code in `doRemoveWatch`:
```
// Skip if there are additional watchers that would prevent us from removing it
m := wm.managedKinds.Get() // Not a deadlock but beware if assumptions change...
if _, ok := m[gvk]; ok {
	return nil
}
```

If you're asking about having multiple watch managers in-process, they would be unaware of each other. I'd say that is an unsupported use-case - should we document that to make it more obvious? Do you have another idea?

Also one more clarification - we no longer need multiple controller-runtime managers in a single process. The parent/child manger relationship previously used has been removed.

maxsmythe replied:
The controller runtime manager and the watch manager share the same cache right?

What happens if I add a sync for constraint templates and then remove it. Is the watch that the controller manager established preserved somehow?

shomron replied:
In practice they wouldn't affect each other as the watch manger is only dealing in unstructured informers  and the constrainttemplate controller would be using a structured informer. These are completely separate in the InformerCache.
I could think of safeguards to put in but perhaps we should handle those as followup tasks?

maxsmythe replied:
Ah, I see the unstructured/structured non-interaction thing.

I think safeguards would be good as this seems like a very subtle behavioral difference to rely on and I wonder if the fact that they are different caches is officially part of the contract.

That being said, I'm fine with deferring more explicit safeguards to a follow-on change. Hopefully sooner rather than later.

maxsmythe commented on pkg/controller/constraint/constraint_controller.go:
```diff
@@ -47,72 +47,81 @@ var (
 		return err
 	}
 
-	r := newReconciler(mgr, gvk, a.Opa, cs, reporter, a.ConstraintsCache)
```
controller switch will remain necessary until we can get rid of the garbage collection for byPod status (controller must be stopped to avoid contention with cleaning loop).

shomron replied:
Got it - added controller switch back in the latest commit.

maxsmythe commented on pkg/controller/constraint/constraint_controller.go:
```diff
@@ -136,16 +143,24 @@ type ReconcileConstraint struct {
+		log.Error(err, "invalid constraint GroupVersion", "gvk", gvk)
 		return reconcile.Result{}, nil
 	}
+	// TODO(OREN) do we need to ask the watch manager if we are registered for this kind?
```
Asking the watch manager seems superfluous, as the channel wouldn't have received the event w/o registration.

shomron replied:
A scenario I was considering was when events are in-flight in the various queues and are received after we have deregistered the watch. I think concretely for this controller it will be fine - de-registration happens after the parent template is removed from OPA, by which point either constraints should have been removed by the garbage collector or if their cache lags, a lookup in OPA should fail and we should drop the request.

maxsmythe commented on pkg/controller/sync/sync_controller.go:
```diff
@@ -120,10 +120,11 @@ func (r *ReconcileSync) Reconcile(request reconcile.Request) (reconcile.Result,
 		return reconcile.Result{}, err
 	}
 	// For some reason 'Status' objects corresponding to rejection messages are being pushed
```
nit: delete comment too

maxsmythe commented on pkg/dynamiccache/cache.go:
```diff
@@ -0,0 +1,66 @@
+	// Construct a new Mapper if unset
+	if opts.Mapper == nil {
+		var err error
+		opts.Mapper, err = apiutil.NewDiscoveryRESTMapper(config)
```
We should be sure to use the dynamic rest mapper here so that we can handle newly created resources.

Question: is it ever possible for the rest mapper to become stale (i.e. kind or resource name gets re-used)? I'm guessing not as the resource -> gvk mapping (and vice versa) should be stable, right?

shomron replied:
^ this code path was not being hit as the manager sets the dynamic rest mapper [here](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/manager/manager.go#L359) and then passes it to the cache [here](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/manager/manager.go#L238-L249).

But yes this is a bug waiting to happen, I'll change it.

Regarding your question, I believe the mapping is stable as well. What might change over time is version priority? Not sure this affects us.

shomron replied:
DynamicRESTMapper is now set explicitly.

maxsmythe commented on pkg/dynamiccache/internal/informers_map.go:
```diff
@@ -0,0 +1,342 @@
+/*
```
How hard do you think it will be to maintain this fork?

Does it make sense to try to upstream this?

shomron replied:
The changes are not pervasive so it's not untenable to maintain, but no fork is always better :)

I can definitely pursue it upstream, but I wouldn't want that to block out efforts here.

maxsmythe commented on pkg/watch/manager.go:
```diff
@@ -33,330 +36,268 @@ type Manager struct {
-	var kindStr []string
-	for gvk := range kinds {
-		kindStr = append(kindStr, gvk.String())
+	if _, ok := wm.watchedKinds[gvk]; !ok {
```
Do we care if we are not watching? Or should this be silent b/c mission accomplished.

shomron replied:
Agreed.

maxsmythe commented on pkg/watch/manager.go:
```diff
@@ -33,330 +36,268 @@ type Manager struct {
-			}
-		}
-		log.Info("previously watched resources have gone pending, removing them from the watch list", "pending", missing)
+	if _, ok := wm.watchedKinds[gvk]; ok {
```
Here is where we could run the cache list and broadcast events to the newly-interested controller.

shomron replied:
Things get tricky when joining an active informer. A few considerations:
1. `AddWatch` needs to be idempotent, we'd need to differentiate if the watcher is actually new. Not much of a problem.
2. Events can be coming in concurrently to the List() for replay. The informer interface doesn't let us pause its internal processing, so here's some wild ideas:

* We might pause receiving/dispatching events from that particular informer on our end until the replay concludes. We may need to do some max-resource-version based dropping of watch events once we resume, as they will have already reached the informer store.
* We might rework things to add a separate event handler per-registrar. `AddEventHandler` properly joins the stream within the informer. When removing the registrar, we wouldn't be able to remove the handler so we'd need to blackhole any events destined for it.
* We might keep track of incoming watch events during the replay. We could drop replay events from List() if their resourceVersion is older than a watch event we observed.

I like the second or the third the best, what do you think? Or did I overcomplicate this?

Note we only have to worry when multiple registrars watch the same resource kind, aka constraints.

maxsmythe replied:
Agreed on idempotence.

Do concurrent updates during List matter?

My understanding is that the work queue debounces pending/processing requests, and that all we know is the Namespace/Name (and now GVK) of the resource, so there is no worry about conflicting data. The controller grabs the data from the server as it processes the event.

It seems like if the above paragraph is true, then the worst effect of updates + list is the potential for over-reconciling by one iteration per resource.

shomron replied:
You're absolutely correct, we don't care about the discrete events and throw away most of the object state in the event in EventPacker in favor of a reference. 👍

maxsmythe commented:
This looks extremely promising! I may have missed some notes b/c its late, but overall this should really simplify things and get us 90% of the way to removing the need for finalizers. The last step being wiping the constraints cached in OPA when a template is deleted.

shomron commented:
> The last step being wiping the constraints cached in OPA when a template is deleted.

...which I believe we were going to address within the [constraint framework](https://github.com/open-policy-agent/frameworks/tree/master/constraint), correct?

shomron commented:
> The last step being wiping the constraints cached in OPA when a template is deleted.

...which I believe we were going to address within the [constraint framework](https://github.com/open-policy-agent/frameworks/tree/master/constraint), correct?

ritazh commented on pkg/controller/constraint/constraint_controller.go:
```diff
@@ -47,72 +46,81 @@ var (
-	finalizerName = "finalizers.gatekeeper.sh/constraint"
+	finalizerName      = "finalizers.gatekeeper.sh/constraint"
+	constraintsGroup   = "constraints.gatekeeper.sh"
+	constraintsVersion = "v1beta1"
```
the version can change so let's not hardcode this

shomron replied:
Good callout 👍

ritazh commented on pkg/controller/config/config_controller.go:
```diff
@@ -85,10 +87,21 @@ func (a *Adder) InjectWatchManager(wm *watch.Manager) {
-	syncAdder := syncc.Adder{Opa: opa}
+	// Events will be used to receive events from dynamic watches registered
+	// via the registrar below.
+	events := make(chan event.GenericEvent, 1024)
```
Is there an upper limit to how many sources we can have?

shomron replied:
There is not. Each controller has their own channel to receive event fan-out from the watch manager. The channel is buffered only to provide some cushion in case the controller is slow to consume from that channel. Consumption is comprised of encoding the event and putting it into a work queue, so it is not blocked on reconciliation time and I do not expect this to become an issue.

The current event distribution is a bit naive and can be improved:
```
	// Distribute the event
	for _, w := range watchers {
		select {
		case w <- e:
		// TODO(OREN) add timeout
		case <-stop:
		}
	}
```

^ theoretically a slow consumer that ends up blocking would cause event distribution to block for other consumers as well. This can be improved with a goroutine per consumer in distributeEvent, but at the cost of additional complexity.

shomron commented:
Known open tasks before final review:

- [ ] Controllers will consult with the watch manager and skip processing a ReconcileRequest if its GVK is no longer being watched. This is to enable disabling controllers (which may have events already in their queues) during shutdown when main.go does status teardown.
- [ ] Determine whether a NOTICE file is needed to call out the controller-runtime cache fork.
- [ ] Unit tests for the new functionality

maxsmythe commented on pkg/watch/manager.go:
```diff
@@ -33,330 +38,337 @@ type Manager struct {
+	u.SetGroupVersionKind(gvk)
+	informer, err := wm.mgr.GetCache().GetInformer(u)
+	if err != nil || informer == nil {
+		// This is expected to fail if a CRD is unregistered.
```
What do we do if Add is called too early for the resource? Return the error and wait on retry? This could cause some controllers to stop working properly if they get one bad resource (such as a sync config on a resource that doesn't exist). It may be worth keeping the old model here, where Add() just creates the intent and a separate goroutine on a timer scans for established resources and actually updates the informers.

This also makes a call to add non-blocking, which should make the config and ConstraintTemplate controllers this would also allow some parallelism in bootstrapping watches at startup

shomron replied:
> What do we do if Add is called too early for the resource? Return the error and wait on retry? This could cause some controllers to stop working properly if they get one bad resource (such as a sync config on a resource that doesn't exist). 
That was the plan. I don't _think_ that could disable a controller, as controller-runtime uses a  rate limiting work queue with per-item rate limiting - other requests should get a chance while this one goes into backoff.

There is another way things could get stuck - if a gvk is known but then `List` fails continuously - `GetInformer()` does a `WaitForCacheSync()`. It sounds like even with the previous model, I'd still need to run `GetInformer()` in a separate goroutine.

+1 on non-blocking and adding parallelism for startup.

shomron replied:
@maxsmythe I started sketching out how a reconcile-based approach [might look](https://github.com/shomron/gatekeeper/blob/dynamic-cache-reconcile/pkg/watch/manager.go#L198-L326), and wasn't happy with how complex it was turning out. Ultimately I opted for adding a non-blocking variant of `GetInformer` called  [`GetInformerNonBlocking`](https://github.com/shomron/gatekeeper/blob/3142f18f09d0fb31a8fdac7315868b7de4c692bb/pkg/dynamiccache/informer_cache.go#L144-L155). Please take a look and let me know if that addresses your concerns above.

shomron commented:
New stuff!

* Reintroduced the controller switch. We had discussed an option of using the watch manager as a switch, however that would not address reconciles mid-way through racing status updates with the main teardown logic.
* The config controller now passes a filtering OpaData client to the sync controller. This client will drop any resources that are not in the current watch set as specified by the config controller, which addresses any potential ReconcileRequests till in-flight when the watch set changes.
* Tests :) Fixed tests in the various controllers. Added unit and integration tests for the watch manager. Still would like a few more (to test replay).
* Tests :( Investigating test failures in the forked dynamic cache. I believe this may be a version mismatch (forked from 0.5.0, but our dependencies are still 0.4.0).

shomron commented:
@maxsmythe this should be ready for review.

maxsmythe commented on pkg/watch/manager.go:
```diff
@@ -1,362 +1,380 @@
-	watchedKinds map[schema.GroupVersionKind]vitals
-	cfg          *rest.Config
-	newDiscovery func(*rest.Config) (Discovery, error)
+	watchedMux   sync.Mutex
```
nit: move comment next to variable

maxsmythe commented on pkg/watch/manager.go:
```diff
@@ -1,362 +1,380 @@
+	switch {
+	case len(watchers.registrars) > 0:
+		// Someone else was watching, replay events in the cache to the new watcher.
+		if err := wm.replayEvents(r, gvk); err != nil {
```
`replayEvents()` appears to be a blocking call that will wait for the informer to warm its cache due to the call to `List()`, or am I missing something?

Could we run this in a separate goroutine? We would need some sort of context that we could cancel if the resource became unwatched while the list was still replaying. Or we could just rely on the receiver to filter out things it's not interested in, like we are doing with the sync controller. The context solution seems better as it avoids an explosion of goroutines if there is a flapping watch/no-watch cycle.

Flapping is a risk we should consider, some kind of debouncing may be necessary to avoid repeated List() calls.

shomron replied:
I've implemented something along these lines in the latest push, please take a look!
Goroutines are contained - there will only be a single per-registrar-per-gvk while replaying events.

Regarding flapping: List calls will be against the informer cache, not the API Server. Assuming there is another registrar watching the same gvk (which is necessary to trigger the replay logic) - a second registrar repeatedly unregistering and registering should not be an issue.

If we are concerned against the primary (first) registrar unregistering and re-registering, we could add rate limiting in the future I'd propose considering that as followup work.

maxsmythe commented on pkg/watch/manager_integration_test.go:
```diff
@@ -0,0 +1,542 @@
+	"testing"
+	"time"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
```
nit: consolidate import blocks

maxsmythe commented on pkg/watch/manager_integration_test.go:
```diff
@@ -0,0 +1,542 @@
+loop:
+	for {
+		select {
+		case <-ctx.Done():
```
what's the timeout for the manager to exit? Is there a timeout? Otherwise we risk an infinite loop.

shomron replied:
5 seconds, set here:

```
ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```

Beyond that go tests have a timeout of 10 minutes by default:
```
        -timeout d
            If a test binary runs longer than duration d, panic.
            If d is 0, the timeout is disabled.
            The default is 10 minutes (10m).
```

maxsmythe replied:
Missed the context timeout, thanks!

maxsmythe commented on pkg/watch/manager_integration_test.go:
```diff
@@ -0,0 +1,542 @@
+
+// Verify ReplaceWatch functionality.
+func TestRegistrar_Replace(t *testing.T) {
+	// TODO
```
Should we test this given the sync controller uses it?

maxsmythe commented on pkg/watch/manager_suite_test.go:
```diff
@@ -0,0 +1,68 @@
+	"sync"
+	"testing"
+
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
```
nit: consolidate import blocks

maxsmythe commented on pkg/watch/registrar.go:
```diff
@@ -1,9 +1,26 @@
 	"fmt"
 	"sync"
 
+	"sigs.k8s.io/controller-runtime/pkg/event"
```
nit: consolidate import groups

shomron replied:
I've been keeping separation between standard library and external imports. Is that acceptable or do you prefer a flat list mixing both?

maxsmythe replied:
That separation is fine, there are some places where there are more than 2 import blocks though.

shomron replied:
Sorry about that, these are artifacts of GoLand / goimports introducing breaks and me not paying much attention to the collapsed blocks. I've fixed where you called them out.

maxsmythe replied:
ACk, thanks.  I hear you WRT the goimports thing. I get the same thing and try to stay on it, but it's definitely not 100% success rate on my end either.

maxsmythe commented on pkg/watch/manager_test.go:
```diff
@@ -1,454 +1,342 @@
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
```
nit: consolidate import groups

maxsmythe commented on pkg/watch/registrar.go:
```diff
@@ -23,54 +45,54 @@ func (w *vitals) merge(wv vitals) (vitals, error) {
-	var addFns []AddFunction
-	for r := range w.registrars {
-		addFns = append(addFns, r.addFns...)
+func (w *vitals) clone() vitals {
```
Is `clone()` used anywhere?

shomron replied:
It's not. I can remove it.

maxsmythe replied:
Thanks!

shomron replied:
Removed.

shomron commented on pkg/dynamiccache/internal/informers_map.go:
```diff
@@ -0,0 +1,358 @@
+	// informer has been started.
+	startWait chan struct{}
+
+	// createClient knows how to create a client and a list object,
```
Yeah I'd rather minimize the fork changes to make it easier to maintain, and pursue these upstream.

shomron commented on pkg/watch/set.go:
```diff
@@ -0,0 +1,168 @@
+func (w *Set) Union(other *Set) *Set {
+	w.mux.RLock()
+	defer w.mux.RUnlock()
+	other.mux.RLock()
```
Fixed in latest push.

shomron commented:
Notes on latest changes:

* replay events now performed asynchronously
* `doAddWatch`: Only fetch informer for first watcher
* `doRemoveWatch`: remove registrar from watch list even when it is not the last watcher
* `replaceWatches`: fix bug that prevented watches from being added if there was another registrar watching them
* tests: reset controller-runtime metrics registry between tests to avoid duplicate registration errors
* add test for `ReplaceWatch`
* `watch.Set`: remove double-locking

maxsmythe commented on pkg/watch/replay.go:
```diff
@@ -0,0 +1,201 @@
+		case <-ctx.Done():
+			return nil
+		case req := <-wm.replayRequests:
+			byRegistrar, inProgress := m[req.gvk]
```
Should `inProgress` be triggered by whether the replay is in `byRegistrar`?

What happens if Registrar A adds a watch, Registrar B adds a watch (and gets replay), and then Registrar C adds a watch? Will a replay trigger?

shomron replied:
You're right, this looks like a bug. I'll fix and add a corresponding test scenario. Good catch!

shomron replied:
Extended test (which initially failed 👍) and fixed.

maxsmythe commented on pkg/watch/manager_test.go:
```diff
@@ -1,454 +1,644 @@
+}
+
+// Verify ReplaceWatch replaces the set of watched resources for a registrar. New watches will be added,
+// unneeded watches will be removed, and watches that haven't change will remain unchanged.
```
nit: change -> changed

shomron replied:
Fixed.

maxsmythe commented:
One issue: were we going to add a NOTICE file?

shomron commented:
@maxsmythe above I wrote ..._Will followup with comprehensive NOTICE in a separate PR. The fork is documented and original copyright is maintained._

Reading through the [Apache guidelines](http://www.apache.org/dev/licensing-howto.html#step-by-step) - it's not clear to me that it's necessary.

They say:

> NOTICE is reserved for a certain subset of legally required notifications which are not satisfied by either the text of LICENSE or the presence of licensing information embedded within the bundled dependency. Aside from Apache-licensed dependencies which supply NOTICE files of their own, it is uncommon for a dependency to require additions to NOTICE.

> Copyright notifications which have been relocated from source files (rather than removed) must be preserved in NOTICE. However, elements such as the copyright notifications embedded within BSD and MIT licenses need not be duplicated in NOTICE -- it suffices to leave those notices in their original locations.

... and in bold:

> Do not add anything to NOTICE which is not legally required.

[controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) does not include a NOTICE file we would duplicate. We include their original license and give attribution to the fork.

I'm happy to follow your guidance on this and/or address this at a later time with legal consultation.

WDYT?

shomron commented:
@maxsmythe ok, I've moved the controller-runtime forked code under `third_party`, added that to various Dockerfiles and Makefiles, marked modified code inline with a disclaimer, brought in the original controller-runtime LICENSE file, and added a new top-level NOTICE file. I also threw in a .dockerignore for good measure. Let me know how this looks.

maxsmythe commented:
LGTM

ritazh commented on pkg/controller/constraint/constraint_controller.go:
```diff
@@ -47,85 +46,103 @@ var (
 	constraintsCache *ConstraintsCache) reconcile.Reconciler {
 	return &ReconcileConstraint{
-		Client:           mgr.GetClient(),
+		// Separate reader and writer because manager's default client bypasses the cache for unstructured resources.
```
> Separate reader and writer because manager's default client bypasses the cache for unstructured resources.

Is there any documentations around this?

could there be any issues with `reader.Get` or `reader.List` when reading from the cache? e.g. data freshness

shomron replied:
I don't know that it's documented, but the code is here: https://github.com/kubernetes-sigs/controller-runtime/blob/v0.5.0/pkg/manager/manager.go#L328-L330

Regarding data freshness - we should be good. Although the watch.Manager does not block for cache sync when adding a new watch, if you do a `reader.Get` or `reader.List`, controller-runtime _will_ block on `WaitForCacheSync` here: https://github.com/open-policy-agent/gatekeeper/blob/5ff95d9bee9aa424f532974891b4385609afde9d/third_party/sigs.k8s.io/controller-runtime/pkg/dynamiccache/internal/informers_map.go#L203

ritazh commented on pkg/controller/sync/sync_controller.go:
```diff
@@ -39,56 +37,53 @@ import (
-	client.Client
-	cs     *watch.ControllerSwitch
+	reader client.Reader
+
```
whats the reason for removing this for sync?

shomron replied:
The ControllerSwitch was only needed in controllers that would fight conflict with the teardown logic in `main.go:L194`. The sync controller isn't writing status or removing finalizers so there shouldn't be any contention to worry about.

shomron commented:
@maxsmythe / @sozercan / @ritazh thank you all for your time and thoughtful reviews! Are there any remaining concerns, or are we good to merge?

maxsmythe commented:
Mergeable from my view. I can merge after Rita's LGTM
